### PR TITLE
Implement database unit of work

### PR DESCRIPTION
**Pull request by xUser5000**
# Problem
A request should create an ACID transaction, that should be committed only when the request finishes successfully. If a request fails (for whatever reason: be it 4xx or 5xx error), the transaction should rollback, effectively preventing any partial writes (e.g., insert queries that span multiple entries or tables) to persist in the database.

## Helpful links (mostly with typeorm)
- https://medium.com/@dev.muhammet.ozen/advanced-transaction-management-with-nestjs-typeorm-43a839363491
- https://stackoverflow.com/questions/77849015/how-to-use-a-transaction-across-multiple-services-in-nestjs
- https://github.com/LuanMaik/nestjs-unit-of-work
- https://github.com/LuanMaik/nestjs-typeorm-transaction-unitOfWork-AsyncLocalStorage
- https://medium.com/@rahulrulz680/unit-of-work-in-repository-pattern-b48a862d06a2

# Solution
- Setup an interceptor (`TransactionInterceptor`) that
  - creates a drizzle transaction upon each request,
  - injects the transaction instance into GraphQL context object (which is shared globally and can be used by any request-scoped provider),
  - calls the route handler,
  - rolls back the transaction if the handler throws an exception,
  - otherwise, commits the transaction.
- Make `DrizzleService` [request-scoped](https://docs.nestjs.com/fundamentals/injection-scopes), which allows it to access GraphQL context object, and consequently access the transaction instance that was created and injected by `TransactionInterceptor`.
- `DrrizleService` should then expose that transaction instance to any provider that wants to use the service.

# Side effects
- Since `DrizzleService` is request-scoped, any provider that depends on it will also be request-scoped (even if it wasn't explicitly decorated with `@Injectable({ scope: Scope.REQUEST })`) [[sauce](https://docs.nestjs.com/fundamentals/injection-scopes#scope-hierarchy)]. Also, because it's a core service that almost all resolvers/services are going to use, that means almost all resolvers/services in our app are going to be request-scoped. This will have a slight performance overhead because request-scoped providers are instantiated every time a request is made to our API and then garbage-collected when the request is finished. This is in contrast to normal providers, which are instantiated once throughout the lifetime of our Nest app. [[sauce](https://docs.nestjs.com/fundamentals/injection-scopes#performance)]
- There's another side effect that stems from the above point, which is the fact that testing now has to be done a little differently (as demonstrated by the changes in `api/src/app.resolver.spec.ts`). This is something I don't fully understand (yet), but here's the [[sauce](https://docs.nestjs.com/fundamentals/testing#testing-request-scoped-instances)] for further reading.

### CODE REVIEW

**Comment by G0maa** [comment]
@xUser5000 
Put sources I sent you in the PR.

File: apps/api/src/app.module.ts [hunk 1]
```diff
@@ -19,6 +19,9 @@ import { AuthModule } from "./modules/auth/auth.module";
             playground: true,
             sortSchema: true,
             autoSchemaFile: join(process.cwd(), "schema.graphql"),
+            // https://github.com/nestjs/nest/issues/1905#issuecomment-479431252
+            // @ts-ignore
```
Comment (G0maa) [inline apps/api/src/app.module.ts:23]: Use `@ts-expect-error: <reason>`. It's better for when they eventually (?) fix the issue.

File: apps/api/src/drizzle/drizzle.service.ts [hunk 2]
```diff
@@ -1,14 +1,21 @@
...
 import * as schema from "./schema";
+import { DRIZZLE_DATABASE_KEY } from "./transaction.interceptor";
 
-@Injectable()
+@Injectable({ scope: Scope.REQUEST })
 export class DrizzleService {
-    public db: NodePgDatabase<typeof schema>;
+    constructor(
+        @Inject(CONNECTION_POOL) private readonly pool: Pool,
+        @Inject(CONTEXT) private readonly context: any,
```
Comment (G0maa) [inline apps/api/src/drizzle/drizzle.service.ts:13]: Can't this by typed? :thinking:

File: apps/api/src/drizzle/transaction.interceptor.ts [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+import { Observable, from, throwError } from "rxjs";
+import { catchError, switchMap } from "rxjs/operators";
+import { CONNECTION_POOL } from "./drizzle.module-definition";
+import * as schema from "./schema";
+
+export const DRIZZLE_DATABASE_KEY = "DRIZZLE_DATABASE";
+
+@Injectable()
+export class TransactionInterceptor implements NestInterceptor {
+    constructor(@Inject(CONNECTION_POOL) private readonly pool: Pool) {}
```
Comment (G0maa) [inline apps/api/src/drizzle/transaction.interceptor.ts:20]: Errm... iirc I think we don't inject db this way :thinking:

Comment (G0maa) [reply apps/api/src/drizzle/transaction.interceptor.ts:20]: hmm
I think I'm missing something here @xUser5000

**Comment by G0maa** [comment]
3ash ya5oya :rocket: :fire: :man_dancing:

File: apps/api/src/app.module.ts (hunk 1 above)
Comment (xUser5000) [reply apps/api/src/app.module.ts:23]: ok

File: apps/api/src/drizzle/drizzle.service.ts (hunk 2 above)
Comment (xUser5000) [reply apps/api/src/drizzle/drizzle.service.ts:13]: will look into it

File: apps/api/src/drizzle/transaction.interceptor.ts (hunk 3 above)
Comment (xUser5000) [reply apps/api/src/drizzle/transaction.interceptor.ts:20]: Yes, I'm aware that this is not the normal way to inject the drizzle service.
The problem if I do it in the normal way (`private readonly drizzleService: DrizzleService`) is that the transaction interceptor now depends on the drizzle service, which in turn exposes something (drizzle db/transaction instance) that should be set by the transaction interceptor itself... circular dependency

In order to resolve this cycle, I had to inject the same primitive (the connection pool) used by the drizzle service to create the db/transaction instance.

Comment (xUser5000) [reply apps/api/src/drizzle/transaction.interceptor.ts:20]: Although, you pointed out something that I overlooked. I don't really know if creating multiple `drizzle()` instances can cause some resource leaks.
I'll investigate that

Comment (G0maa) [reply apps/api/src/drizzle/transaction.interceptor.ts:20]: @xUser5000 
btw, will using `AsyncLocalStorage` help in regards to eliminating the need for `request scoped` _everything_?

Comment (xUser5000) [reply apps/api/src/drizzle/transaction.interceptor.ts:20]: @G0maa 
I asked claud 3.5 sonnet about the code and whether or not it has resource leaks. It gave me this:

---
To evaluate potential resource leaks in this code, let's analyze the key components:

Connection Pool:
The Pool from the pg library is created in the DrizzleModule and injected into the DrizzleService. Pools are designed to manage and reuse connections efficiently, which helps prevent leaks.
DrizzleService Scope:
The DrizzleService is scoped to REQUEST, meaning a new instance is created for each request. This doesn't directly cause leaks, but it's important to ensure resources are properly managed.
Drizzle ORM Usage:
The drizzle function is called in the db getter of DrizzleService. It either returns an existing database instance from the context or creates a new one.

Potential Concerns:

Multiple Drizzle Instances:
If the db getter is called multiple times within a request and DRIZZLE_DATABASE_KEY is not set in the context, it could create multiple drizzle instances. However, this doesn't necessarily lead to leaks as long as the underlying Pool is managed correctly.
Transaction Management:
The presence of TransactionInterceptor suggests that transactions are being managed, which is good for resource handling. However, we'd need to see its implementation to ensure it's correctly releasing resources.
Pool Shutdown:
There's no visible code for shutting down the Pool when the application terminates. This could potentially lead to connection leaks if not handled elsewhere.

Conclusion:
Based on the provided code, there are no immediate, obvious resource leaks. However, to ensure complete resource safety:

Ensure the Pool is properly closed when the application shuts down.
Verify that the TransactionInterceptor correctly manages database transactions and releases resources.
Consider caching the drizzle instance in the DrizzleService to avoid creating multiple instances per request.

To be absolutely certain, you'd need to test the application under load and monitor for any connection or memory leaks. Also, reviewing the implementation of TransactionInterceptor and any application shutdown hooks would provide a more complete picture.
---

So, in conclusion, I think we're safe when it comes to database connection leaks or anything like that.

Comment (xUser5000) [reply apps/api/src/drizzle/transaction.interceptor.ts:20]: Regaring `AsyncLocalStorage`, you are right, we can potentially throw away request-scoped providers and replace and make them access the request object through `AsyncLocalStorage`.

I'll investigate that further and update the PR for another review

Comment (xUser5000) [reply apps/api/src/drizzle/transaction.interceptor.ts:20]: Investigation results: spent 3 hours trying to make `AsyncLocalStorage` work alongside the `GraphQLModule` and wasn't lucky enough ig

However, Claude suggested another solution to mitigate the problem of all providers being request-scoped. In a nutshell, we can create a "proxy" provider around the drizzle service that is not request-scoped and does a little bit of magic to retrieve a correct instance of the drizzle service. I'll try to look into it.

**Comment by G0maa** [comment]
@xUser5000 
I'll merge, but:
1. I think there are lots of stuff that need to be documented in the Wiki from the comments on this PR / PR itself.
2. About ALS :thinking:, thing is Node.js is "single-threaded" in the terms that every request does not create it's own thread, this (I think?) fixes many problems introduced by the thread-per-request scheme, such as database pooling. But produces problems such as "context" per request.

The thing about services, they do not have access to "context"... welp usually, unless we do it this way.

ALS should provide us with that. It's basically a storage for a request's context.

**Comment by xUser5000** [comment]
@G0maa 
https://github.com/activecourses/disworse/wiki/Database-unit-of-work
